src/Components/Server/test/Circuits/CircuitRegistryTest.cs

```diff
@@ -313,16 +314,108 @@ public async Task DisconnectWhenAConnectIsInProgress()
             Assert.False(registry.DisconnectedCircuits.TryGetValue(circuitHost.CircuitId, out _));
         }
 
+        [Fact]
+        public void CircuitRegistryUsesConfiguredMaxRetainedDisconnectedCircuitsValue()
+        {
+            // Arrange
+            var circuitIdFactory = TestCircuitIdFactory.CreateTestFactory();
+            var maxCircuits = 3;
+            var circuitOptions = new CircuitOptions
+            {
+                MaxRetainedDisconnectedCircuits = maxCircuits,
+            };
+            var registry = new TestCircuitRegistry(circuitIdFactory, circuitOptions);
+            var hosts = Enumerable.Range(0, maxCircuits + 2)
+                .Select(_ => TestCircuitHost.Create())
+                .ToArray();
+
+            // Act
+            for (var i = 0; i < hosts.Length; i++)
+            {
+                registry.RegisterDisconnectedCircuit(hosts[i]);
+            }
+
+            // Assert
+            for (var i = 0; i < maxCircuits; i++)
+            {
+                Assert.True(registry.DisconnectedCircuits.TryGetValue(hosts[i].CircuitId, out var _));
+            }
+
+            // Additional circuits do not get registered.
+            Assert.False(registry.DisconnectedCircuits.TryGetValue(hosts[maxCircuits].CircuitId, out var _));
+            Assert.False(registry.DisconnectedCircuits.TryGetValue(hosts[maxCircuits + 1].CircuitId, out var _));
+        }
+
+        [Fact]
+        public async Task DisconnectedCircuitIsRemovedAfterConfiguredTimeout()
+        {
+            // Arrange
+            var circuitIdFactory = TestCircuitIdFactory.CreateTestFactory();
+            var circuitOptions = new CircuitOptions
+            {
+                DisconnectedCircuitRetentionPeriod = TimeSpan.FromSeconds(3),
+            };
+            var registry = new TestCircuitRegistry(circuitIdFactory, circuitOptions);
+            var mre = new ManualResetEventSlim();
+
+            registry.OnAfterEntryEvicted = () =>
+            {
+                mre.Set();
+            };
+            var circuitHost = TestCircuitHost.Create();
+
+            registry.RegisterDisconnectedCircuit(circuitHost);
+
+            // Act
+            // Verify it's present in the dictionary.
+            Assert.True(registry.DisconnectedCircuits.TryGetValue(circuitHost.CircuitId, out var _));
+            await Task.Run(() => Assert.True(mre.Wait(TimeSpan.FromSeconds(10))));
```

Comment:
Is there a more reliable way to do this other than using an mre. Maybe simple set a flag in OnAfterEntryEvicted and do await Task.Delay(4) before checking the flag was set?

Reply:
Out of interest, why would that be more reliable? I would have thought the mre was preferable and more reliable, because it continues immediately after the flag is set, so we can give it up to 10 seconds in case code on one thread happens to run much faster than code on another thread.

Reply:
I meant simpler. Reliable is probably not the right way. I'm still waking up :)

Reply:
I remember @Tratcher changing all MRE's recently to TCS's for reliability. I'll make that change and await the TCS.Task instead here.